src/Microsoft.Extensions.Configuration.AzureAppConfiguration/FeatureManagement/FeatureManagementKeyValueAdapter.cs

```diff
@@ -247,6 +240,995 @@ public bool NeedsRefresh()
+                            {
+                                featureVariant.StatusOverride = reader.GetString();
+                            }
+                            else if (reader.TokenType != JsonTokenType.Null)
+                            {
+                                throw CreateFeatureFlagFormatException(
+                                    FeatureManagementConstants.StatusOverride,
+                                    settingKey,
+                                    reader.TokenType.ToString(),
+                                    JsonTokenType.String.ToString());
+                            }
+
+                            break;
+                        }
+
+                    default:
+                        reader.Skip();
+
+                        break;
+                }
+            }
+
+            return featureVariant;
+        }
+
+        private FeatureTelemetry ParseFeatureTelemetry(ref Utf8JsonReader reader, string settingKey)
+        {
+            var featureTelemetry = new FeatureTelemetry();
+
+            while (reader.Read() && reader.TokenType != JsonTokenType.EndObject)
+            {
+                if (reader.TokenType != JsonTokenType.PropertyName)
+                {
+                    continue;
+                }
+
+                string telemetryPropertyName = reader.GetString();
+
+                switch (telemetryPropertyName)
+                {
+                    case FeatureManagementConstants.Enabled:
+                        {
+                            if (reader.Read() && (reader.TokenType == JsonTokenType.False || reader.TokenType == JsonTokenType.True))
+                            {
+                                featureTelemetry.Enabled = reader.GetBoolean();
+                            }
+                            else if (reader.TokenType == JsonTokenType.String && bool.TryParse(reader.GetString(), out bool enabled))
+                            {
+                                featureTelemetry.Enabled = enabled;
+                            }
+                            else if (reader.TokenType != JsonTokenType.Null)
+                            {
+                                throw CreateFeatureFlagFormatException(
+                                    FeatureManagementConstants.Enabled,
+                                    settingKey,
+                                    reader.TokenType.ToString(),
+                                    $"{JsonTokenType.True}' or '{JsonTokenType.False}");
+                            }
+
+                            break;
+                        }
+
+                    case FeatureManagementConstants.Metadata:
+                        {
+                            if (reader.Read() && reader.TokenType == JsonTokenType.StartObject)
+                            {
+                                featureTelemetry.Metadata = new Dictionary<string, string>();
```

Comment:
Metadata looks like a good candidate for parsing using `JsonDocument.ParseValue`

Reply:
Maybe, but from what I saw, telemetry metadata should be a dictionary with string keys and string values, and not any json object. Is there a way to validate after it gets parsed that it fits that structure?

Reply:
Good point